### PR TITLE
Ann's BESTUFF heroku app

### PR DESCRIPTION
**Pull request by gofarann**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a custom model method you wrote. | I wrote one to determine whether the user has already voted for the work called already_voted? in the work model. it takes in a user as a parameter and sees if there are records in the votes db where the work id and user id already exists. the method returns true or false. |
| Describe how you approached testing that model method. What edge cases did you come up with? | I approached testing, poorly. I had fixtures missing parameters. but otherwise pretty pitiful. I need to keep working on the tests.   |
| Describe an edge case test you wrote for a controller | see above. |
| What are `session` and `flash`? What is the difference between them? | session is the controller that redirects based on whether someone is logged in or not. Flash is the notice that notifies when an action errors or succeeds. |
| Describe a controller filter you wrote. | I'm not sure what a controller filter is.  |
| What was one thing that you gained more clarity on through this assignment? | well.. I did not understand nested routes until the latter half of the project. So none of my work category controllers should have been there. I learned more about that later on in the upvote method.  |
| What is the Heroku URL of your deployed application | https://bestuff.herokuapp.com/ |
| Do you have any recommendations on how we could improve this project for the next cohort? | Overall, I loved how challenging this was. Very challenging. This one as a group project would have been good because I could have learned for folks with better knowledge of how the entire thing worked. I think the CSS grid framework portion could have been a different project. |

### CODE REVIEW

File: test/models/work_test.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+require "test_helper"
+
```
Comment (CheezItMan) [inline test/models/work_test.rb:2]: What about tests for the custom methods?

File: app/views/works/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+</section>
+</main>
+
+
+<!-- could have put this in a method, not sure where -->
+
+<% top_movies = @movies.sort { |a, b| a.votes.count <=> b.votes.count}.take(10).reverse %>
+<% top_books = @books.sort { |a, b| a.votes.count <=> b.votes.count}.take(10).reverse %>
+<% top_albums = @albums.sort { |a, b| a.votes.count <=> b.votes.count}.take(10).reverse %>
+
```
Comment (CheezItMan) [inline app/views/works/index.html.erb:19]: This should probably be in the Models, not in the view or the controller.

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    <title>MediaRanker</title>
+    <%= csrf_meta_tags %>
+
+    <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
+
+    <nav class="header">
+      <h2 class="title"><%= link_to "Media Ranker", root_path %><small> Ranking the Best of Everything</small></h2>
+
+    <nav class="login">
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:13]: This `nav` block should be in the `body` element.

File: test/models/vote_test.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  it "should be created with user and work ids" do
+    vote = votes(:my_vote)
+    vote.valid?.must_equal true
+  end
+
+  it "should not be created with no user id" do
+    vote = votes(:my_vote_no_user_id)
+    vote.valid?.must_equal false
+  end
+
```
Comment (CheezItMan) [inline test/models/vote_test.rb:15]: You should also test relationships.

File: test/controllers/books_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+   it "should redirect to book list after adding book" do
+     post works_path, params: { work:
+     { category: "book",
+       title: work.title
+       }
+     }
+     must_redirect_to books_path
+
+   end
+
```
Comment (CheezItMan) [inline test/controllers/books_controller_test.rb:25]: In these controller tests you should also be testing negative cases.

**Review by CheezItMan — COMMENTED** [review]
Nicely done, you've got some work to go on the testing, but overall a nice job.  Check out the notes i put into the code.

**Comment by CheezItMan** [comment]
# Media Ranker
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | Lots of commits, good commit messages and it looks like you had branches, but I can't see them in Github.   |
| Comprehension questions	| "session is the controller that redirects based on whether someone is logged in or not. Flash is the notice that notifies when an action errors or succeeds,"  `Session` is actually a hash like object that stores information that can be accessed in the controller.   |
| **General** |  |
| Rails fundamentals (RESTful routing, use of named paths) | Check |
| Semantic HTML | Check |
| Errors are reported to the user | Good use of `flash` to let user know about problems, and successful upvotes/login. |
| Business logic lives in the models | Good use `top_vote` and `already_voted?`    |
| Model testing | A few notes were added |
| Controller testing | Check, you do need to check testing negative cases. |
| **Wave 1 - Media** |  |
| Splash page shows the three media categories | Check |
| Basic CRUD operations on media are present and functional | Check |
| **Wave 2 - Users and Votes** |  |
| Users can log in and log out | Check |
| The ID of the current user is stored in the session | Check |
| Individual user pages and the user list are present | Check |
| A user cannot vote for the same media more than once | Check |
| All media lists are ordered by vote count | The category page aren't ordering the items, but the homepage is. |
| Splash page contains a media spotlight | Check |
| Media pages contain lists of voting users | Check  |
| **Wave 3 - Styling** |  |
| Foundation is used appropriately | Well done! |
| Look and feel is similar to the original | Check |
|  **Overall** | Nicely done, you've got some work to go on the testing, but overall a nice job.  Check out the notes i put into the code.   |
